### DISCUSSION TITLE
Required Properties limitations

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#7134] [Language Ideas]
Can you please allow private and protected required properties?  I would like to use this for static factory methods.

**Comment by DavidArno** [comment]
What would be the use-case for private required properties? I'm unsure what scenario this would cover that a private constructor can't cover, eg:
```cs
public class C 
{
    private C() {}
    
    public required int P { get; init; }
    
    public static C Factory(int i) => new C { P = i };
}

var x = C.Factory(1); // compiles
var y = new C { P = 1 }; // doesn't compile
```

**Reply by TonyValenti** [reply]
@DavidArno - everything with required properties could formerly have been done with constructors. 

But required properties are much nicer because of the object initializer syntax.

**Reply by RikkiGibson** [reply]
It would be worthwhile to review the required members working group notes and the LDM notes on this feature. Basically, for your suggestion to happen, whatever reasons we had for blocking less-accessible required properties would need to be revisited and addressed.

**Reply by TonyValenti** [reply]
Digging in a little bit, this is where the situation was discussed:

meetings/2022/LDM-2022-01-05.md

And it seems as if the scenario was never really considered. I believe these rules should be revised.  A property should be able to be required and the scope around that should be controlled by the call site and not the definition of the property.  Simply stated, if a public class has a protected required property, the constructor should only be able to be called in locations where the required property can be set. The fact that it is not publicly settable should not prohibit me from using it the way I want.

**Reply by HaloFour** [reply]
So what if the public class has a public constructor but protected or internal required properties?  What does that even mean?  That's why this wasn't considered, it makes the feature a lot more complicated for very little gain.  You're no longer talking about the public surface of a type as it is intended to be consumed by an external assembly, you're talking about guidelines internal to your own assembly, which means substantially less ROI.

**Reply by RikkiGibson** [reply]
I guess I could imagine some rule where only constructors with SetsRequiredMembers can be more accessible than the least accessible required property. But the enforcement feels wonky. Whose fault is it if the accessibility doesn't align? When the containing type is the limiting factor (required members can't be less accessible than it), it's easier to say that the member is at fault for not aligning with its containing type.

**Reply by TonyValenti** [reply]
I don't think there is an alignment issue at all.

Conceptually, required properties are an alternative syntax for constructors. 

A public class can have a private constructor. Why can't a public class have a private required properties?  They are different syntaxes for essentially the same thing.

**Reply by DavidArno** [reply]
> So what if the public class has a public constructor but protected or internal required properties? What does that even mean? 

Currently, I can declare a public class, with a public constructor that sets all of the public required properties, but the compiler will complain if the constructor is used on its own. I have to mark that constructor with `[SetsRequiredMembers]` to reassure it that those required properties are managed correctly. 

If the public class has a public constructor and non-public required properties, then the same would apply: that constructor would have to be annotated with `[SetsRequiredMembers]`. There is nothing special here about those properties being non-public.

> That's why this wasn't considered, 

I find that a very non-compelling justification for why it might not have been considered.

At the moment, required properties are very much an MVP feature. They do enough of the job to be useful. But they are limited and ungainly. I'm sure in time, the team will return to look at improving this feature. At that point, looking at supporting non-public required properties would make complete sense. Of course, the team might still conclude it being too niche a feature to worry about, but they might also decide it worth it. Time may tell...

**Reply by HaloFour** [reply]
@DavidArno 

> If the public class has a public constructor and non-public required properties, then the same would apply: that constructor would have to be annotated with `[SetsRequiredMembers]`. There is nothing special here about those properties being non-public.

Which would defeat the purpose of declaring them as required properties. You'd either have a type you could never construct because you can't assign the non-public required properties, or you'd have a type that has required properties which are never enforced.  It solves no use cases to make the language behave in that way.

> I find that a very non-compelling justification for why it might not have been considered.

Seems like the best reason to me.  Don't chase all of the rabbits down all of the holes for the sake of doing so.  Wait for feedback that demonstrates use cases where expanding upon the feature in those more complicated scenarios will provide value and has behavior that makes sense.

I'm not arguing that expanding required properties in these scenarios shouldn't be considered, only that doing so requires collecting enough data about the potential use cases to inform what the behavior should be, if it's worth doing at all.

**Comment by zorgoz** [comment]
This is even weirder: why a declaration-level object initializer does not satisfy the required modifier?

```c#
public class WeirdRequired
{
	public required int[] MyProperty { get; set; /*or init*/ } = new int[] {1,3,4};
}

void Main()
{
	var wr = new WeirdRequired();
}
```
_CS9035 Required member 'UserQuery.WeirdRequired.MyProperty' must be set in the object initializer or attribute constructor._

**Reply by HaloFour** [reply]
The `required` metadata is a part of the public surface of the type, the existence of an initializer is not.  One could argue that a `required` property with an initializer might be worth a warning or other diagnostic, but there might be reasons to want a defensive initializer for scenarios that don't directly support or understand required properties, like deserialization.

**Reply by zorgoz** [reply]
Well, no initializer is "public surface".
I can accept that the semantics of the constructor itself can't be (it could be) evaluated by the compiler to determine if all required are set, and that's why the `SetsRequiredMembersAttribute` was introduced, but these codes are no less prone to what you are saying but still complies. 
`var wr = new WeirdRequired { MyProperty = new int[] {1,3,4} };`
or 
```c#
public class WeirdRequired
{
	public required int[] MyProperty { get; init; }
	[SetsRequiredMembers] // I lied!
	public WeirdRequired()
	{
		
	}
}
```
So nope, I don't think that this is elaborated enough...

**Reply by HaloFour** [reply]
> I can accept that the semantics of the constructor itself can't be (it could be) evaluated by the compiler to determine if all required are set

The compiler doesn't get into the business of attempting to decompile methods to determine what they do.  With reference assemblies, the body of the method will always be blank anyway.

As required properties affect the public surface of a given type the language team explicitly doesn't want it to be inferred by the compiler.  They want the developer to explicitly state which properties are required, and explicitly state when initialization of those required properties is not required for a given constructor.

**Reply by zorgoz** [reply]
This still does not make any difference between "creation-time" and "declaration-time" initializers. Neither has anything to do with publicity or constructors. And both can be assessed at compile-time. The error message itself is ambiguous as both are object initializers.

**Reply by HaloFour** [reply]
@zorgoz 

"Declaration-level initializers" are not a part of the public surface of a type.  They are an implementation detail, code buried in the body of the constructor.  External consumers of your type do not know that they exist.  All a consumer sees is whether or not that property is required, and the compiler enforces that.  The fact that a constructor might also internally set a value isn't relevant, that's not part of the API.

**Reply by zorgoz** [reply]
Well, according to [**this article**](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-11.0/required-members#enforcement), it is the constructor's responsibility. And the constructors are the ones validated to satisfy these criteria. Hence autogenerated constructors could be too. 

But if you take a look here: 
https://sharplab.io/#v2:EYLgtghglgdgPgAQMwAIECYUFEAeEwAOANgKYCwAUAN6Up1qoBOJAjgK5TMAmKsALgG0AuigBqEImxIoqKAOYk+AbhQBnRSoC+KALwoYJAO68YgkVQCMmpbXq26ybHkKkAFAEp7Mr/TESpuvpGJmYyVjYUvpqU0RSUQA

You will notice something here:
![image](https://user-images.githubusercontent.com/1569170/236823865-afd2ebb7-af42-4285-a1d8-cdb726bf2332.png)
The red part is the autogenerated "constructor", and the green is the explicit constructor part (merged into one). If you remove the explicit constructor, the implicit one will still have that value set.
![image](https://user-images.githubusercontent.com/1569170/236824929-df763b17-93ff-4d25-aade-188b636b8daa.png)

IL does the same. And the `[CompilerFeatureRequired("RequiredMembers")]` is attached to both.
Hence: the enforcement is satisfied with the implicit constructor as well.

The only difference is that the implicit code is setting the backing field directly, while the explicit one is setting the property itself. 
Indeed, there could be differences if the property is not auto-implemented. 

And this is the rationale, IMHO: the two kinds of initializers are not implemented in the same way under the hood.

**Reply by HaloFour** [reply]
Either form of initialization from within the class is a part of the implementation details of the class and doesn't affect the public surface.  So if the constructor is not marked as `[SetsRequiredMembers]`, then it's expected that the consumer of that class will be required to set `Value`.  The compiler is smart enough to note that the property-level initializer does satisfy the requirements of `[SetsRequiredMembers]` and does not issue a warning.
